frontend/lib/src/components/widgets/DataFrame/columns/utils.ts

```diff
@@ -633,3 +633,43 @@ export function removeLineBreaks(text: string): string {
   }
   return text
 }
+
+/**
+ * Determines the correct value to display in a link cell based on the `href` and `displayText` parameters.
+ *
+ * @param href - The raw url value.
+ * @param displayText - The display text, which can either be a regex pattern that will be applied the `href` or a string to be displayed.
+ * @returns - The string value to be displayed in the cell.
+ *
+ * * @example
+ * getLinkDisplayValue("https://roadmap.streamlit.app", "https:\/\/(.*?)\.streamlit\.app"); // returns "roadmap"
+ * getLinkDisplayValue("https://roadmap.streamlit.app", undefined); // returns ""https://roadmap.streamlit.app""
+ */
+export function getLinkDisplayValue(
+  href: string,
+  displayText?: string
+): string {
+  if (!displayText) {
+    return href
+  }
+
+  try {
+    // assume displayText is a regex first and attempt to find a pattern match in the url
+    // u flag allows unicode characters
+    // s flag allows . to match newlines
+    const displayTextRegex = new RegExp(displayText, "us")
+
+    // apply the regex pattern to display the value
+    const patternMatch = href.match(displayTextRegex)
+    if (patternMatch && patternMatch[1] !== undefined) {
+      // return the first matching group
+      return patternMatch[1]
+    }
+
+    // if the regex doesn't find a match with the url (or it wasn't a regex pattern at all), return the displayText value
```

Comment:
I played around with this a bit, and unfortunately, I think this behaviour makes this more confusing for the user :(

https://github.com/streamlit/streamlit/assets/2852129/9f9b6518-7966-41c9-9277-b49c7c6b17fb

As soon as there is a mismatch between the input and regex, it will show the regex as display value... which is probably unexpected in many cases for developer and user

Reply:
It would be even better if we could also restrict the click to the height of the text.. but I think this is a lot more complicated, so might not be worth it.

Reply:
I also wonder if we care about optimizing the runtime of display logic. For example, should I opt to just iterate one time through the string instead of doing:
`if (displayText.includes("(") && displayText.includes(")")) {`
?

Reply:
Yeah, I think this part is indeed something to keep quite "optimized" since it is called a lot of times in situations like scrolling. One way on how I have been dealing with similar situations is by running some pre-computations on a column level in the column initialization function (and not within `getCell`). E.g. the `validateRegex` here is constructed at column load and not during cell rendering: 

<img width="650" alt="image" src="https://github.com/streamlit/streamlit/assets/2852129/7e29b1c8-6ae0-48f4-b146-9a3877574ff4">

If you do something similar for display text, you could do something like:

```
let displayText: RegExp | string | undefined = parameters.display_text

if (displayText.includes("(") && displayText.includes(")")) {
    displayText = new RegExp(parameters.validate, "us")
}
```

And in `getCell` you directly use the `getLinkDisplayValue(href, displayText)` and in `getLinkDisplayValue` you check if it is a `displayText instanceof RegExp` -> apply it as regex otherwise use it as display value.

Reply:
so, basically make the decision if its a regex or a plain string during column load, and then only apply it during rendering

Reply:
Nice optimization!